text-document.rkt

```diff
@@ -567,24 +532,57 @@
      (match-define (doc doc-text doc-trace)
        (hash-ref open-docs (string->symbol uri)))
      (define pos (- (line/char->pos doc-text line char) 1))
-     (define range
+     (define-values (start end)
        (match ch
          ["\n"
-          (Range
-           #:start (abs-pos->Pos doc-text (send doc-text paragraph-start-position line))
-           #:end (abs-pos->Pos doc-text (send doc-text paragraph-end-position line)))]
+          (values
+           (abs-pos->Pos doc-text (send doc-text paragraph-start-position line))
+           (abs-pos->Pos doc-text (send doc-text paragraph-end-position line)))]
          [_
-          (Range
-           #:start (abs-pos->Pos doc-text (or (find-containing-paren pos (send doc-text get-text))
-                                              0))
-           #:end (abs-pos->Pos doc-text pos))]))
-
-     (range-formatting!
-      id
-      (hash-set params
-                'range range))]
+          (values
+           (abs-pos->Pos doc-text (or (find-containing-paren pos (send doc-text get-text))
+                                      0))
+           (abs-pos->Pos doc-text pos))]))
+     (success-response id (format! uri start end 'on-type))]
     [_
-     (error-response id INVALID-PARAMS "textDocument/onTypeformatting failed")]))
+     (error-response id INVALID-PARAMS "textDocument/onTypeFormatting failed")]))
+
+;; Shared path for all formatting requests
+;; `kind` allows diverging in a few places while sharing the core operation
+(define (format! uri start end kind)
```

Comment:
we might need a contract to ensure `kind` is valid?

Reply:
Potentially, though it doesn't seem like any of the "internal" functions in this file use contracts at the moment... Maybe it's not needed if we switch to a flag like you suggested below?